sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py

```diff
@@ -489,75 +507,172 @@ def _resolve_inputs_for_pipeline_component_jobs(self, jobs, base_path):
+                node.compute = resolver(node.compute, azureml_type=AzureMLResourceType.COMPUTE)
+            if has_attr_safe(node, "compute_name") and not is_data_binding_expression(node.compute_name):
+                node.compute_name = resolver(node.compute_name, azureml_type=AzureMLResourceType.COMPUTE)
+
+    @classmethod
+    def _divide_nodes_to_resolve_into_layers(
+        cls,
+        component: PipelineComponent,
+        extra_operations: List[Callable]
+    ):
+        """Traverse the pipeline component and divide nodes to resolve into layers.
+        For example, for below pipeline component, assuming that all nodes need to be resolved:
+          A
+         /|\
+        B C D
+        | |
+        E F
+        return value will be:
+        [
+          [("B", B), ("C", C), ("D", D)],
+          [("E", E), ("F", F)],
+        ]
+
+        :param component: The pipeline component to resolve.
+        :type component: PipelineComponent
+        :param extra_operations: Extra operations to apply on nodes during the traversing.
+        :type extra_operations: List[Callable]
+        :return: A list of layers of nodes to resolve.
+        :rtype: List[List[Tuple[str, BaseNode]]]
+        """
+        # add an empty layer to mark the end of the first layer
+        layers, cur_layer_head, cur_layer = [list(component.jobs.items()), []], 0, 0
 
-        for key, job_instance in jobs.items():
-            preprocess_job(job_instance)
+        while cur_layer < len(layers) and cur_layer_head < len(layers[cur_layer]):
+            key, job_instance = layers[cur_layer][cur_layer_head]
+            cur_layer_head += 1
+
+            cls._resolve_binding_on_supported_fields_for_node(job_instance)
             if isinstance(job_instance, LoopNode):
                 job_instance = job_instance.body
-            if isinstance(job_instance, AutoMLJob):
-                self._job_operations._resolve_arm_id_for_automl_job(job_instance, resolver, inside_pipeline=True)
-            elif isinstance(job_instance, BaseNode):
-                resolve_base_node(key, job_instance)
-            elif isinstance(job_instance, ConditionNode):
-                pass
-            else:
-                msg = f"Non supported job type in Pipeline: {type(job_instance)}"
-                raise ComponentException(
-                    message=msg,
-                    target=ErrorTarget.COMPONENT,
-                    no_personal_data_message=msg,
-                    error_category=ErrorCategory.USER_ERROR,
-                )
 
-    def _resolve_arm_id_and_inputs(self, component):
+            for extra_operation in extra_operations:
+                extra_operation(job_instance, key)
+
+            if isinstance(job_instance, BaseNode) and isinstance(job_instance._component, PipelineComponent):
+                if cur_layer + 1 == len(layers):
+                    layers.append([])
+                layers[cur_layer+1].extend(job_instance.component.jobs.items())
```

Comment:
Do we need to skip ConditionNode here?

Reply:
ConditionNode is not a BaseNode so won't be processed here